### ISSUE TITLE
WC5 - get files from Hyrax to S3

### ISSUE THREAD

**Issue by decimalator** [OregonDigital/OD2#526]
### Descriptive summary

We need to figure out what it's going to take to store assets' files in S3.  This means at a minimum the main content file(s) and all derivatives thereof.  I don't believe there are other files we need to send to S3, but I'm not sure.

### Expected behavior

- On ingest, the content stream is written to S3
- Derivatives are generated and then stored in S3
  - These may need to be generated on disk and then copied up to S3
- Downloads of the main assets check user permissions, then stream the file from S3 instead of the file system
- Thumbnails are served directly from S3 so searches aren't horribly slow
- Derivatives for viewers are read from S3
  - These may need to be cached locally depending on a variety of factors.  For instance, RAIS has to cache JP2 files because a single JP2 is typically serving up several tiles, and we don't want each tile to take 1-2 seconds to render.

### Related Issues

#524 - if we implement a fairly small lambda for derivatives, it completely changes how we go about putting those files into S3.  We'll need a decision there before we start any real work here.  (It does *not* affect how we convince Hyrax to **read** those derivatives, however)

### Accessibility Concerns

N/A - this is entirely backend work

**Comment by jechols** [comment]
Research notes will follow in this comment - I'll edit it to keep it updated.  These may need to be reorganized for somebody else to use, but I'll try to make sense as much as I can.

Hyrax::FileSetDerivativesService calls into `derivative_path_for_reference` to determine the "URL" for derivatives, which tells something deeper in the stack where / how to store files (*more research needed here*).  It prefixes the returned paths with "file://" in all cases - this probably needs to be configurable, or maybe `derivative_path_for_reference` should just return a URI instead of a flat string.

Hyrax::DerivativePath is used in a few places as the derivative path-finding service.  It's got a lot of logic that assumes the filesystem is where derivatives live: `Dir.glob` is used in the `all_paths` method, and `Pathname` is used in two places, including as a return value.  `Pathname` might be fine, I'm not sure, but it was built specifically for filesystem paths, so there may be areas it breaks down when using something like `s3://foo/bar/baz.jp2`.

`Hyrax::DownloadsController.load_file`: overrides a hydra-core method.  It gets a "default" file (*more research needed - is this the asset's main file?*) if a "reference" isn't passed in.  If a reference is passed in, it uses that to get a derivative file URL.  From the context, this must mean that "reference" is something like "thumbnail", "jp2", etc.  This function uses `File.exist?`, which means it's hard-coded to look on the filesystem.

**Comment by jechols** [comment]
Mentioned in the meeting, capturing here just in case:

- We still should look into options for externalizing the derivative generation now that we know Hyrax doesn't couple the asset with the location its derivatives live
- Looking on and off into various options for changing out our derivative jobs - I think I'll have to spin up a new ticket for this
- My focus for this ticket will be making Hyrax *read* from the cloud, not write.  We can circle back to what it'll take to write to the cloud if we have to, but external derivative generation would remove the need to do that.

**Comment by jechols** [comment]
So here's where I'm at, and what I need to pick up next time:

- hydra-core has the default download behavior.  It's a gem that actually lives in the hydra-head gem's repo for some reason: https://github.com/samvera/hydra-head/tree/master/hydra-core. Hyrax overrides this and we may need to override hyrax.  blah.
- RAIS will be handling the JP2 derivatives, so we should be able to just spit out an s3 URL and configure RAIS for s3.  This needs to be prototyped.
- Downloads are where we're most likely to see derivatives requested that aren't for RAIS.  I believe this is how thumbnails are presented, too.  (It is in OD1)
- ACK still need to see about the fedora setting(s?) for s3 storage!
  - If fedora isn't easily configurable to read/write s3 we have a bigger investigation - so this needs to stop being put on the back-burner when I get back to this
- Easiest way to see what needs to change at this point: change the return from the core derivative path functions to spit out a fake "s3://blah" url and see what happens.  but this may hose backend stuff, too, that isn't obvious.  So this isn't a catch-all.  Dang.
-

**Comment by jechols** [comment]
So after a lot of investigations, this ticket hasn't really moved forward in any significant way.  Where I'm at now:

- https://github.com/OregonDigital/OD2/tree/feature/deriv-generation-separation contains all the stuff I'm working on.  It's very prototypey still.
- I currently have the system sending jobs to a very slim derivative worker (under 30 megs of RAM vs. about 280 for the standard worker container), and it queues jobs up for hyrax fileset updates.  There are problems:
  - When Hyrax gets the job back from my slim worker, I get an odd error, `NoMethodError: undefined method 'jid=' for #<UpdateFilesetJob:0x000056342c30a988>`.  The stack trace tells me this is sidekiq's doing, so I suspect there's some magic in the way ActiveJob works on the Rails end that I'm not properly emulating when I manually queue a job from the slim derivative worker.
  - There is still no connection to S3 - I'm *still* simply trying to make the jobs external before implementing anything "real"
  - I have been digging around in Hyrax code again, and I think we're still going to have a fair number of places to override just to make Hyrax not care about the files being in S3.  Not as bad as if we keep derivatives in Hyrax, but still bad.
  - We may want to share code in a very simple way so that the slim worker uses the same logic as the main app for determining a derivative's path.  Rails and Hyrax being what they are, it will be tricky to build a standalone piece of logic that can be shared by both apps without making the slim worker get a lot fatter.  Alternatively, we just copy and paste logic, but this leaves us in a position where a mistake can get us out of sync - e.g., the workers put files somewhere Hyrax is no longer referencing.

**Comment by jechols** [comment]
The branch now has changes which enable image derivatives to be created 100% externally, and the app reads them properly.  Still to do:

- It isn't smart yet: JP2s are given exactly 6 resolution levels in all cases rather than based on image size
- Implement PDFs
- Implement audio and video files, neither of which I know much about
- Use something in the environment to decide where to expect derivatives
  - In dev, for instance, we use a local file path.  Easy!
  - In production, we want S3 - the generator needs to put files in S3 and Hyrax needs to produce S3-ified URLs for RAIS
    - This gets even trickier for a/v because Hyrax has to read those files and present them to users, but still respect access controls
    - Hmm... does RAIS need to have access controls to avoid people just sharing a IIIF url?
  - The staging server should use S3 so it mirrors production, but we want it in a different bucket

**Comment by jechols** [comment]
Considering localstack: https://github.com/localstack/localstack to avoid dev vs. prod drift

**Comment by jechols** [comment]
Went with [MinIO](https://min.io/product) instead of localstack.  It only does S3, but it seems to be an actual production-quality server.  localstack was slow to start, slow to shut down, had a lot of weird things going on that made diagnosis of problems very difficult, and persistence of data was simply not feasible.

**Comment by jechols** [comment]
The stack is now able to run a derivative job on images, store the derivative in S3, and view the derivative in RAIS and Hyrax (Hyrax for thumbnails).

Now I'm stuck trying to figure out how to intercept fileset and work deletions to clean up derivatives.  These behaviors seem not to exist in Hyrax in any form.  By replacing the Hyrax core `CleanupFileSetsActor`, I can intercept deletes on works, but it isn't hit for filesets.  I can't seem to implement my own actor for deletes otherwise, despite having copied the subclass, modules, and method name directly from the `CleanupFileSetsActor`.

Other derivative jobs still have yet to be implemented.

**Comment by jechols** [comment]
In digging, @luisgreg99 found the `FileSetActor` which could be overridden to provide derivative deletion.  In that code, there is a wonderful `:on_destroy` callback I was hoping we could hook into in order to deal with this!  Unfortunately, for some reason that makes no sense, `Hyrax.config.callbacks.set` overwrites existing callbacks.  And there is no "append" option.  So we could set our own callback, but then we lose the default behavior.  We could set our own callback and include the functionality they have in their default, but if the default has to change for any reason in some Hyrax update, we won't even know we're missing functionality.

I gotta ask: *why did they bother to expose callbacks in the first place?*

Okay, so maybe we could shim another actor in after the `FileSetActor`, right?  No.  This class, despite being called an "actor", isn't used in the same way as other actors.  There is no middleware-like set of actors where the `FileSetActor` lives.  It is instead instantiated as needed.

So it looks like our only option is indeed monkey-patched overrides.  Again.  Oh, but it's not even that simple.  We can override `FileSetActor.destroy` - but that is only called when a single FileSet is destroyed.  It is *not* called when a work is destroyed.  So we need to also override what I described previously (`CleanupFileSetsActor`).

FFS

**Comment by jechols** [comment]
Actually, there is hope after all!  Looks like FileSet is already overridden locally.  I'm not sure why, since we don't seem to have done anything with it, but there you have it.  And it appears so far that overriding its `destroy` method could be a single point of entry for handling the derivative removals.  Yay!

**Comment by jechols** [comment]
After getting deletion working, I rebased the branch.  We now have some "File Manager" thing instead of direct links to delete images, and it crashes on me, so I'm guessing there's another "images are expected to be local" thing that was updated since I started this.

**Comment by jechols** [comment]
File manager problem was unrelated.

Unfortunately, looking at the PDF setup, I am noticing that PDFs have *two* types of derivatives, and one of them (extracted text) has to be stored in Fedora with very particular attributes.

It might make sense to try and split the PDF derivatives job into two parts: one would still live in our main stack since it only stores files in Fedora, not the filesystem; the other would just replace the local derivative setup.  I'm hoping I don't run into this sort of thing with other derivative types, but I bet I do.  Especially with the "document" types.

**Comment by jechols** [comment]
Time to close this and start a new ticket or two... or three....

The external derivatives are working for images and PDFs, but the code really needs to live in its own code repository.  This will definitely complicate the stack, as well as dev and maintenance, but if we want to change how we handle external derivatives, a separate repo is easier.  And none of the derivative stack belongs in OD2's codebase.
